### ISSUE TITLE
Implement sub-shell entry/exit instead of just environment diffing

### ISSUE THREAD

**Issue by dmayle** [direnv/direnv#798]
**Is your feature request related to a problem? Please describe.**
I use a Linux system that doesn't conform to FHS (NixOS), which for most things has no effect on what I do.  When I'm developing software, however, there are times when the (downloaded) tools I use expect FHS.  My specific instance was bazel, which bootstraps its own toolchain, downloading code to compile proto or go files, for example.

For NixOS, there is a tool that allows you to enter a sub-shell with a chroot into an FHS environment.  For most of my development needs, I use direnv to jump in and out of configurations, but since direnv only uses the sub-shell for building up the env diff, the chroot is lost, so I have to keep track of which environments require FHS, and manually enter those.

**Describe the solution you'd like**
This works, but it's not ideal.  When looking a bit through the direnv source and hook, I see that the direnv hook eval's the direnv output in order to setup the environment.  It would be possible to have the eval enter a sub-shell instead, setup to return an 'exit' directory on process completion.

In the sub-shell would be a modified direnv hook.  When exiting the direnv directory, the hook would exit the sub-shell, returning the new-current directory, allowing the external shell to properly update.

**Describe alternatives you've considered**
I've considered just updating .envrc to call a sub-shell instead, and print a message to let me know that I've entered a subshell, and need to exit the subshell manually.   I've also considered just living with the solution as it is.

**Comment by zimbatm** [comment]
I get what you are saying and it also crossed my mind. Another advantage of sub-shell is that they can also support loading shell aliases and functions. The `.envrc` format would have to be shell agnostic so that it's understood by all shells. This would be essentially a different tool.

**Comment by zimbatm** [comment]
In some ways, that would be a simpler and more generic tool.

One difficulty would be to integrate with editors if they are started outside of the project. With direnv, it's possible to propagate the env diff to the editor.

**Comment by dmayle** [comment]
@zimbatm I don't think it would be a good idea to make the change for all projects.  I think that the direnv approach is much simpler.  I was thinking about adding something to the .envrc (like `export DIRENV_SUBSHELL=1`) that would change the behavior for just that environment.

This would enable a mode just for those projects that require the sub-shell.

**Comment by rmchale** [comment]
I will give my vote for this.  It would help python virtualenv development

**Comment by hexagonrecursion** [comment]
Would this also let me automatically enter/exit toolbox? Ref: https://github.com/direnv/direnv/issues/890

**Comment by Oudwins** [comment]
Has there been any progress on fixing this?
